glslc/test/working_directory.py

```diff
@@ -0,0 +1,210 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os.path
+
+import expect
+from environment import File, Directory
+from glslc_test_framework import inside_glslc_testsuite
+from placeholder import FileShader
+
+
+@inside_glslc_testsuite('WorkDir')
+class TestWorkDirNoArg(expect.ErrorMessage):
+    """Tests -working-directory."""
+
+    glslc_args = ['-working-directory']
+    expected_error = [
+        "glslc: error: argument to '-working-directory' is missing "
+        '(expected 1 value)\n',
+        'glslc: error: no input files\n']
+
+
+@inside_glslc_testsuite('WorkDir')
+class TestWorkDirEqNoArg(expect.ErrorMessage):
+    """Tests -working-directory=<empty>."""
+
+    glslc_args = ['-working-directory=']
+    expected_error = ['glslc: error: no input files\n']
+
+
+@inside_glslc_testsuite('WorkDir')
+class TestWorkDirEqNoArgCompileFile(
+    expect.SuccessfulReturn, expect.CorrectObjectFilePreamble):
+    """Tests -working-directory=<empty> when compiling input file."""
+
+    environment = Directory('.', [
+        Directory('subdir', [File('shader.vert', 'void main() {}')])
+    ])
+    glslc_args = ['-c', '-working-directory=', 'subdir/shader.vert']
+
+    def check_object_file_preamble(self, status):
+        return self.verify_object_file_preamble(
+            # output file should be generated into subdir/.
+            os.path.join(status.directory, 'subdir/shader.vert.spv'))
+
+
+@inside_glslc_testsuite('WorkDir')
+class TestWorkDirEqNoArgLinkFile(
+    expect.SuccessfulReturn, expect.CorrectObjectFilePreamble):
+    """Tests -working-directory=<empty> when linking input file."""
+
+    environment = Directory('.', [
+        Directory('subdir', [File('shader.vert', 'void main() {}')])
+    ])
+    glslc_args = ['-working-directory=', 'subdir/shader.vert']
+
+    def check_object_file_preamble(self, status):
+        return self.verify_object_file_preamble(
+            # output file should be generated into the current directory.
+            os.path.join(status.directory, 'a.spv'))
+
+
+@inside_glslc_testsuite('WorkDir')
+class TestWorkDirCompileFile(
+    expect.SuccessfulReturn, expect.CorrectObjectFilePreamble):
+    """Tests -working-directory=<dir> when compiling input file."""
+
+    environment = Directory('.', [
+        Directory('subdir', [File('shader.vert', 'void main() {}')])
+    ])
+    glslc_args = ['-c', '-working-directory=subdir', 'shader.vert']
+
+    def check_object_file_preamble(self, status):
```

Comment:
This could go into a base class (together with `environment`), parameterized by directory.